src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs

```diff
@@ -0,0 +1,850 @@
+        internal const int c_MaxWriteDataSize = 63 * 1024; // 1k for the framing and trailer that is always less as per SSPI.
+
+        private bool _canRetryAuthentication;
+        private ProtectionLevel _expectedProtectionLevel;
+        private TokenImpersonationLevel _expectedImpersonationLevel;
+        private uint _writeSequenceNumber;
+        private uint _readSequenceNumber;
+
+        private ExtendedProtectionPolicy _extendedProtectionPolicy;
+
+        // SSPI does not send a server ack on successful auth.
+        // This is a state variable used to gracefully handle auth confirmation.
+        private bool _RemoteOk = false;
+
+        internal NegoState(Stream innerStream, bool leaveStreamOpen)
+        {
+            if (innerStream == null)
+            {
+                throw new ArgumentNullException("stream");
+            }
+
+            _innerStream = innerStream;
+            _leaveStreamOpen = leaveStreamOpen;
+        }
+
+        internal static string DefaultPackage
+        {
+            get
+            {
+                return NegotiationInfoClass.Negotiate;
+            }
+        }
+
+        internal void ValidateCreateContext(string package,
+                                            NetworkCredential credential,
+                                            string servicePrincipalName,
+                                            ExtendedProtectionPolicy policy,
+                                            ProtectionLevel protectionLevel,
+                                            TokenImpersonationLevel impersonationLevel)
+        {
+            if (policy != null)
+            {
+                // One of these must be set if EP is turned on
+                if (policy.CustomChannelBinding == null && policy.CustomServiceNames == null)
+                {
+                    throw new ArgumentException(SR.net_auth_must_specify_extended_protection_scheme, "policy");
+                }
+
+                _extendedProtectionPolicy = policy;
+            }
+            else
+            {
+                _extendedProtectionPolicy = new ExtendedProtectionPolicy(PolicyEnforcement.Never);
+            }
+
+            ValidateCreateContext(package, true, credential, servicePrincipalName, _extendedProtectionPolicy.CustomChannelBinding, protectionLevel, impersonationLevel);
+        }
+
+        internal void ValidateCreateContext(
+                                            string package,
+                                            bool isServer,
+                                            NetworkCredential credential,
+                                            string servicePrincipalName,
+                                            ChannelBinding channelBinding,
+                                            ProtectionLevel protectionLevel,
+                                            TokenImpersonationLevel impersonationLevel
+                                            )
```

Comment:
Why are these parameters indented so much?  Usually only one tab stop of indent is used.